File: app/src/main/java/com/example/elsafa/ui/Product/ProductPresenter.java
package com.example.elsafa.ui.Product;

import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;

import Model.Product.CompleteProduct;

public class ProductPresenter {

    private final OnGetProductListener listener;

    public ProductPresenter(OnGetProductListener listener) {
        this.listener = listener;
    }

    private void getProductFromFirebase(String categoryName, String productId) {
        FirebaseFirestore fStore = FirebaseFirestore.getInstance();

        fStore.collection("Categories")
                .document(categoryName)
                .collection("Products")
                .document(productId)
                .get()
                .addOnSuccessListener(new OnSuccessListener<DocumentSnapshot>() {
                    @Override
                    public void onSuccess(DocumentSnapshot ds) {
                        CompleteProduct product = ds.toObject(CompleteProduct.class);
                        product.setProductId(ds.getId());
                        listener.onGetProductSuccess(product);
                    }
                });
    }

    public void getProductInfo(String categoryName, String productID) {
        getProductFromFirebase(categoryName, productID);
    }
}

File: app/src/main/java/com/whiteside/cafe/Product/ProductPresenter.kt
package com.whiteside.cafe.Product

import Model.Item
import Model.Product
import com.google.firebase.firestore.FirebaseFirestore

class ProductPresenter {
    var listener: OnGetProductListener? = null
    fun setListener(listener: OnGetProductListener?) {
        this.listener = listener
    }

    private fun getProductFromFirebase(item: Item?) {
        val fStore = FirebaseFirestore.getInstance()
        fStore.collection("Categories")
            .document(item.getCategoryName())
            .collection("Products")
            .document(item.getProductId())
            .get()
            .addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    val ds = task.result
                    val product = ds.toObject(Product::class.java)
                    product.setProductId(ds.id)
                    listener.onGetProductSuccess(product)
                } else {
                    listener.onGetProductFailed(task.exception)
                }
            }
    }

    fun getProduct(item: Item?) {
        getProductFromFirebase(item)
    }

    fun getProductInfo(categoryName: String?, productId: String?) {
        val item = Item()
        item.categoryName = categoryName
        item.productId = productId
        getProduct(item)
    }
}